### PR TITLE
System receptionist RFC

### PR DESCRIPTION
**Pull request by davidkpiano**
[View rendered RFC](https://github.com/statelyai/rfcs/blob/38a4f804b912f225696b6fe30eb7bba50bc4ae74/0002-system-receptionist.md)

### CODE REVIEW

**Comment by mellson** [comment]
I think the proposal looks good 👍🏻 

You ask if `{ system }` should always be defined when a machine is created.
Personally, I would prefer that it's always there to avoid too much defensive programming at the call site.

Would it be an idea to add a sendTo method for the receptionist? Such that you could do something like this:

```ts
system.receptionist.sendTo({ type: 'complete' }); // default send to all actors
system.receptionist.sendTo({ type: 'complete' }, { firstOnly:true }); // only send to first actor
```
Thanks for the RFC; I look forward to following along here.

**Comment by yaffol** [comment]
> Passing thought.. should you be able to specify tags when registering an actor? That would allow a subscription channel kind of pattern..

This rings true of the kind of ways I automate in AWS/terraform/ansible etc - use of tags is a very flexible way to make flexible but specific groups of things. I can see this being useful in defining and accessing actors/services.

**Comment by davidkpiano** [comment]
> I think the proposal looks good 👍🏻
> 
> You ask if `{ system }` should always be defined when a machine is created. Personally, I would prefer that it's always there to avoid too much defensive programming at the call site.

I agree with this. The `system` should always be defined.

> Would it be an idea to add a sendTo method for the receptionist? Such that you could do something like this:
> 
> ```ts
> system.receptionist.sendTo('todo', { type: 'complete' }); // default send to all actors
> system.receptionist.sendTo('todo', { type: 'complete' }, { firstOnly:true }); // only send to first actor
> ```

This seems like it'll be a common pattern; currently, this is pretty boilerplate-like:

```js
actions: sendTo((ctx, e, { system }) => system.receptionist.find('todos'), { type: 'some.event' })
```

We can probably do something like this:

```js
import { sendToRegistered } from 'xstate/system'; // bikeshed me

// ... 

actions: sendToRegistered('todos', { type: 'some.event' })
```

**Comment by davidkpiano** [comment]
@aceslick911 

> Instead of “find” I think it would be great to have some kind of subscription which would automatically maintain some kind of object in the context…. A hash map would make this a bit faster to access as well.

What do you think of the above pseudocode?

```js
actions: sendToRegistered('todos', { type: 'some.event' })
```

Having a subscription that listens for actor updates and saves the ref to local `context` is just duplicating efforts IMO - there's already a reference to it in the `system` object, so there doesn't seem to be much gained from having an additional reference (I could be wrong)

**Comment by davidkpiano** [comment]
> Passing thought.. should you be able to specify tags when registering an actor? That would allow a subscription channel kind of pattern..

That would be interesting! What would a good API for that look like? Maybe this?

```js
register: {
  key: 'todo',
  tags: ['item']
}

// getting all actors of a specific tag
system.receptionist.getTagged('item') // bikeshed me
```

**Comment by mellson** [comment]
@johtso great idea with tags, and I think the suggested API looks promising.
> ```ts
> register: {
>  key: 'todo',
>  tags: ['item']
> }
> ```
Akka has a similar concept called [ActorTags](https://doc.akka.io/api/akka/current/akka/actor/typed/ActorTags$.html). Let's reuse their idea of having the tags in a set.

**Comment by mellson** [comment]
The suggestion to pull out the most common patterns in some helpers is a great idea.

> We can probably do something like this:
> ```ts
> import { sendToRegistered } from 'xstate/system'; // bikeshed me
> 
> // ... 
> 
> actions: sendToRegistered('todos', { type: 'some.event' })
> ```

And, since you're brave enough to invite some bikeshedding, here we go 😊 

Should the receptionist specific helpers have their own namespace?
```ts
 import { sendToRegistered } from 'xstate/system/receptionist'
 ```
Maybe it's a bit too much, but at least the intent is apparent.

And if we are thinking in terms of small helpers, then maybe two for sending?
```ts
 import { sendToAll, sendToAny } from 'xstate/system/receptionist'
 ```

And lastly, we could consider making the send helpers able to take one or more ids:
```ts
 import { sendToAll } from 'xstate/system/receptionist'

// ...
actions: sendToAll('todos', { type: 'some.event' })
// ...
actions: sendToAll(['todos','todonts','supervisors'], { type: 'some.event' })
 ```

**Comment by mdpratt** [comment]
Our systems uses a receptionist system as well, but implements the tagging system in such a way that you can send events blindly to all actors who match your tag filtering criteria.

For example, our receptionist would have the following (bikeshedding our implementation to match XState paradigms):

```typescript
// Match any machine with this tag
sendToTagged('tagString', { /* Event */ });

// Match any machine with all tags
sendToTagged(['tagA', 'tagB'], { /* Event */ });

// Matches any machines that pass the validation function
const tagMatcher = (tags: string[]): boolean => { /* Logic to match */ };
sendToTagged(tagMatcher , { /* Event */ });
```

Here is a toy example of how this could be used. Here, the credentials provider automatically expires after some time, and when it does, it notifies _all_ machines who are consumers.

```typescript
// Machines that manage UI
const renderGraphWidget = createMachine({ tags: ['credentials-consumer', 'display-graphs'] });
const renderListWidget = createMachine({ tags: ['credentials-consumer', 'display-list'] });

// Higher level manager
const credentialsProvider = createMachine({
  tags: ['credentials-provider']
}, {
  actions: {
    notifyConsumersOfExpired: sendToTagged(
      'credentials-consumer',
      { type: 'CREDENTIALS_EXPIRED'}
    )
  }
});
```

In this toy example, the credentialsProvider machine doesn't care who the machines are, and there is no need to managed callers in context, it will happily blindly notify anyone who's a `credentials-consumer`.

**Comment by xylophonehero** [comment]
I'm very much a fan of this system.

Often at times, I find myself having to make a choice between two methods when setting a parent child relationship between react components which both have machine logic controlling them. The methods are:
1. Interpreting a new machine inside a child component and subscribing to updates from a parent service
2. Spawning a machine inside the parent service and passing the actor down as a prop or accessing it on a snapshot of the parent service in the child component

Both have advantages where the other has disadvantages.
 
Method 1 allows me to set up the actions and services for the machine from inside the component. This is crucial to properly execute side effects (especially Dom manipulation methods for elements in child component) and not triggering them from useEffects that are reacting to state/context changes from the machine. But I can't respond or listen to events happening in the parent machine (where maybe an external service defined in that machine is recieving a stream of events).

Method 2 allows me to communicate effectively with the parent machine. Mostly by directly sending events via the helpful forwardTo/sendParent methods XState provides. But when I call `useActor` in the child component, I can't specify the config options anymore because the machine has already been interpreted. This means to execute a side effect (without `useEffect`), I have to awkwardly set a callback function in the context when the component mounts. And then I can call it when the event occurs directly from the machine. This involves quite a bit of boilerplate code especially when there are multiple callbacks.

The recipienist pattern allows me to use method 1 of interpreting machines while still being able to easily communicate with other machines via message passing. 

Also with this pattern, my machines are no longer bound by a hierarchy of parent and child. My machines can communicate freely with any other machine, subscribing to as many events from other machines as possible, rather than having a single parent which spawns/invokes it that it can communicate easily with.

Will add some code examples later on, writing all this from a phone 😅

File: text/0000-system-receptionist.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+https://doc.akka.io/docs/akka/current/typed/actor-discovery.html
+
+Naturally, actors form a hierarchical relationship, such that it is easiest for a parent actor to communicate with a child actor, and for a child actor to _sometimes_ communicate with the parent, if it has knowledge of it. However, these communication patterns are more difficult:
+
+- An actor communicating with a sibling actor
+- An actor communicating with any arbitrary actor
+- A system-level actor (such as a logging or notification actor) being made available to any actor
+- Retrieving non-stale data from another actor
+- Grouping actors
+- Communicating with groups of related actors
```
Comment (Andarist) [inline text/0000-system-receptionist.md:27]: Is "grouping actors" different from "communicating with groups of related actors"? should those be merged into a single item? if not - could we clarify what's the other use case for grouping actors?
```suggestion
- Communicating with groups of related actors
```

File: text/0000-system-receptionist.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+      actions: (ctx, e, { system }) => {
+        // Lists all todo ActorRefs
+        console.log(system.receptionist.list('todo'));
+      }
+    },
+    'todos.completeAll': {
+      actions: pure((ctx, e, { system }) => {
+        return system.receptionist.list('todo').map(todoRef => {
+          return sendTo(todoRef, { type: 'complete' });
+        });
```
Comment (Andarist) [inline text/0000-system-receptionist.md:102]: unregistering would be easier if we'd keep it internally as a Set and I don't think there is a strong reason to do an implicit conversion to an array on behalf of the caller

File: text/0000-system-receptionist.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+A common use-case of the receptionist will be to register and find a single `ActorRef` registered to a given `key`. This can be done via `system.registry.find(key)`, which will return `ActorRef | null`.
+
+```ts
+const todoMachine = createMachine({
+  // ...
+  on: {
+    complete: {
+      // `system.registry.find(key)` returns 0 or 1 ActorRef objects
+      // TODO: ensure `sendTo(...)` handles null targets (dead-letter queue?)
```
Comment (Andarist) [inline text/0000-system-receptionist.md:126]: how often does such a swallowing behavior implicitly allows bugs and how often this is actually what a developer wants?

File: text/0000-system-receptionist.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    }
+  }
+});
+```
+
+This should be roughly equivalent to:
+```ts
+// PSEUDOCODE
+// system.receptionist.get('notifier');
+system.receptionist.getSnapshot().registry.get('notifier')?.[0] ?? undefined; // => ActorRef[]
```
Comment (Andarist) [inline text/0000-system-receptionist.md:140]: What if there are more actors registered under the requested key? should we throw an error then?

File: text/0000-system-receptionist.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+  listener(key) {
+    // Return ActorRef that listens for registry updates on that key
+    // ...
+  }
+}
+```
+
+
+**Unregistering**
```
Comment (Andarist) [inline text/0000-system-receptionist.md:183]: We should also automatically clean up actors in the receptionist when an actors stops

File: text/0000-system-receptionist.md [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+## Alternatives
+
+TODO
+
+## Unresolved questions
+
+- System options: `createSystem(machine, options?)` (what should go here?)
+- Default system? I.e. should `{ system }` always be defined when a machine is created without `createSystem(...)`?
+  - `createMachine(...)` can be thought of as `createSystem(...)` without any shared resources
+- Need to further explain the role of `createSystem`, especially how it can also solve other common problems such as specifying system-wide implementations to spawned actors without using factory functions
```
Comment (Andarist) [inline text/0000-system-receptionist.md:255]: how does the `interpret` relate to `createSystem`? or should we replace interpreter~ with the system at the API surface level?

File: text/0000-system-receptionist.md [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+
+To mitigate this, we should inform devs that XState can be used successfully without systems (orphaned actors by default) nor receptionists, and that the purpose of these two concepts is to _simplify_ common use-cases that are otherwise difficult and complicated to implement.
+
+## Alternatives
+
+TODO
+
+## Unresolved questions
+
+- System options: `createSystem(machine, options?)` (what should go here?)
```
Comment (Andarist) [inline text/0000-system-receptionist.md:252]: it would be even fine if we wouldn't have any options for it - we just need to keep the second argument slot free for potential options in the future

File: text/0000-system-receptionist.md (hunk 4 above)
Comment (davidkpiano) [reply text/0000-system-receptionist.md:140]: I'm actually not sure, will have to check the behavior of what Akka does; it's not clear in the documentation.

My gut feeling is this should behave just like `array.find(...)`, where it returns the first result regardless of if there are more results.

File: text/0000-system-receptionist.md (hunk 2 above)
Comment (davidkpiano) [reply text/0000-system-receptionist.md:102]: How would you do the "send this event to all todo actors" use-case, then?

**Comment by davidkpiano** [comment]
> Obligatory question - how do we type this? 😉 Especially typing "dynamic" systems with per-system receptionists seem to be quite tricky. A global receptionist would be way easier to type.

One idea is to use that 2nd argument slot to provide the system schema:

```ts
const system = createSystem(todosMachine, {
  schema: {
    registry: {
      notifier: {} as NotifierActorRef,
      logger: {} as LoggerActorRef,
      todo: {} as TodoActorRef
    }
  }
});
```

And for machines that need to "inherit" these registered actors, `system.createMachine(...)` can be used:

```ts
const todoMachine = system.createMachine({
  // has notifier, logger, todo registry actor types from system
});
```

-----

An alternative for this API to make this more clear/organized would be to _not_ pass in a root machine to the system:

```js
const system = createSystem({
  schema: {
    registry: { ... } // as above
  },
  // potentially provide system-wide implementations
  actions: {
    notify: (ctx, e) => { ... }
  }
});

const todosMachine = system.createMachine({
  // has registry actor types and system-wide implementation types
});

const todoMachine = system.createMachine({
  // ...
});

const todosActor = todosMachine.start();
```

**Comment by christoph-fricke** [comment]
I generally like the idea of a receptionist and agree that it can be useful to architect actor-based applications.

However, the current proposal feels largely coupled to actor that are spawned from state machines, given that large part of the receptionist interaction is done through actions.

What about "lightweight" actors that are spawned from the more general interface: `Behavior`?
Not every stateful unit in a system contains enough finite state to justify a state-machine behavior. For some, implementing `Behavior` would be just fine.

For example: An [event-bus](https://github.com/christoph-fricke/xsystem/blob/main/src/messaging/event_bus.ts#L58) or [websocket connection](https://github.com/christoph-fricke/xsystem/blob/main/src/messaging/websocket.ts#L29) can be useful in a system, and should often be accessible globally through the receptionist. But their implementation does not necessarily require a state machine.

**Will it be possible to support `Behavior`-based actors in the receptionist, i.e. provide API's that work at a lower level than state machines?**

In other words: I think an implementation of this RFC should ensure that it is based on `Behavior`s - the lowest interface required for spawning actors. Conceptually, state machine definitions are a more specialized `Behavior`. Ideally, their support (and maybe some helper function, e.g. actions creators) would be the result of the general `Behavior` support.

**Comment by davidkpiano** [comment]
Thanks so much for the feedback!

> What about "lightweight" actors that are spawned from the more general interface: `Behavior`? Not every stateful unit in a system contains enough finite state to justify a state-machine behavior. For some, implementing `Behavior` would be just fine.

I agree. State machines are only one type of behavior implementation. The API can be tweaked to allow for anything spawned/invoked to be registered:

```js
actions: assign((ctx, e, { spawn }) => {
  return spawn('anything', {
    register: 'thing', // register this actor to the 'thing' key
    name: 'thing-1'
  });
}
```

For consistency, this can be combined with the suggestion to make registering a top-level configuration option:

```js
const todoMachine = createMachine({
  register: 'todo',
  // ...
})
```

> For example: An [event-bus](https://github.com/christoph-fricke/xsystem/blob/main/src/messaging/event_bus.ts#L58) or [websocket connection](https://github.com/christoph-fricke/xsystem/blob/main/src/messaging/websocket.ts#L29) can be useful in a system, and should often be accessible globally through the receptionist. But their implementation does not necessarily require a state machine.
> 
> **Will it be possible to support `Behavior`-based actors in the receptionist, i.e. provide API's that work at a lower level than state machines?**
> 
> In other words: I think an implementation of this RFC should ensure that it is based on `Behavior`s - the lowest interface required for spawning actors. Conceptually, state machine definitions are a more specialized `Behavior`. Ideally, their support (and maybe some helper function, e.g. actions creators) would be the result of the general `Behavior` support.

**Review by davidkpiano — COMMENTED** [review]
Update how actors are registered

File: text/0000-system-receptionist.md [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+  listener(key) {
+    // Return ActorRef that listens for registry updates on that key
+    // ...
+  }
+}
+```
+
+
+**Unregistering**
```
Comment (Andarist) [inline text/0000-system-receptionist.md:182]: I have no idea what to do about it yet but it's worth noting that a code like this isn't guaranteed to work in the React world:
```ts
// one component
useMachine({
  entry: register('todo')
})

// another component
useSelector(system.registry.find('todo'))
```

The problem is with the timing of the `register` - we can't easily call it within the render function and thus we might not be able to read from that registered key in another part of the app.

Perhaps the upcoming cache API could help us to solve this but there are not a lot of details available out there about this so I'm not sure.

Comment (davidkpiano) [reply text/0000-system-receptionist.md:182]: @Andarist See the latest changes, where registering is done by the _parent_ at the time of actor creation, rather than as an action on the child:


```ts
import { register } from 'xstate/actions';
const todoMachine = createMachine({
  // Register actor to the 'todo' key
  entry: register('todo'),
  // ...
});
```

This should be roughly equivalent to:
```ts
// PSEUDOCODE
entry: (ctx, e, { system, self }) => {
  system.register('todo', self);
}
```

Comment (davidkpiano) [reply text/0000-system-receptionist.md:185]:
```ts
const todoMachine = createMachine({
  // ...
});

const todosMachine = createMachine({
  on: {
    'todo.create': {
      actions: assign({
        newTodo: (ctx, e, { spawn }) => {
          return spawn(todoMachine, {
            id: `todo-${/* unique id */}`,
            register: 'todo' // multiple actors can be registered here
          })
        }
      })
    }
  }
})
```

Latest changes; registering now occurs at time of creation

Comment (Andarist) [reply text/0000-system-receptionist.md:185]: that doesn't change anything for the case like this:
```ts
createMachine({
  context: ({ spawn }) => spawn(todoMachine, { register: 'todo' })
})
```

The problem is still the same - the registry can't be easily mutated during render and thus it can't be reliably read during the same render by another component.

Comment (davidkpiano) [reply text/0000-system-receptionist.md:185]: In the above case, the spawned machine is available immediately; that is, it's part of `machine.getInitialState(...)` and doesn't need to wait to be started in an effect, since the instance will be the same (just not started yet).

Comment (Andarist) [reply text/0000-system-receptionist.md:185]: Yes, this part is OK - the problem is with the "reading" part here:
```ts
useSelector(system.registry.find('todo'))
```
We can't reliably make the `'todo'` key available during the same render without introducing potential memory leaks in the React model. Some libraries that try to solve this are using **global** timers and/or `FinalizationRegistry`... but that's a giant hack in itself.

The problem is that the "key owner" (`todosMachine` in your example) might not end up mounting despite being rendered at least once - in such a case the key itself should become unavailable in the registry... but how do you cleanup the key added in render?

Comment (davidkpiano) [reply text/0000-system-receptionist.md:185]: The `'todo'` key would be available. When the system is started (machine is started), `system.registry.find('todo')` is immediately available.

Are you approaching this from the concern of components _externally_ registering actors onto the registry? This is not supported/mentioned in this RFC, as only actors within the system can register other actors.

Comment (Andarist) [reply text/0000-system-receptionist.md:185]:
```ts
const Child = () => {
  // the key might not be available here yet
  // even though its parent component "registers" this key at the creation time
  // and not when the machine in the parent starts
  useSelector(system.registry.find('todo')) 
}

const Parent = () => {
  // not that this key wants to conceptually be available asap as it's registered at creation time
  // if we make it available immediately in the registry (that is a shared resource) this would be equivalent to a side-effect in render
  // it's not possible, in a clean, predictable, way to remove such a key in case the `<Parent/>` doesn't get committed by React
  createMachine({
    context: ({ spawn }) => spawn(todoMachine, { register: 'todo' })
  })
  return <Child/>
}
```

Comment (davidkpiano) [reply text/0000-system-receptionist.md:185]: There's a slight misunderstanding: `system` is not "global".

This would look more like this:

```js
const Child = ({ mySystem }) => {
  // Available
  useSelector(mySystem.registry.find('todo'));
}

const Parent = () => {
  const mySystem = useInterpret(createSystem(...)); // system is just an actor

  return <Child mySystem={mySystem} />
}
```

Comment (Andarist) [reply text/0000-system-receptionist.md:185]: It doesn't matter if the system is truly global or if it's scoped to a React subtree - it's still a shared resource.

There is still a scenario in which the system actor gets created and committed while a machine registering into it from within its subtree won't do the same. Imagine a conditional rendering within the subtree, you might render `ChildA` (registering `todo-a`) but before it commits you can switch to `ChildB` (registering `todo-b`). In such a case, it's not obvious how to clean the `todo-a` key from the system

Comment (davidkpiano) [reply text/0000-system-receptionist.md:185]: With the latest changes to this RFC, a machine _never_ registers itself to a system. It's the parent machine that registers child machines to the system.

Contrived example: adding a todo actor (via sending a `'todo.create'` or similar event to the parent actor) would do these things:

1. Add it to the machine context via `spawn(...)`
2. Simultaneously register it with the `system` that the parent actor belongs to
3. **React will rerender**
  - At this point, that todo actor exists.

Note that React is not involved _at all_ until after the actor is registered.
